icicle_v3/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,667 @@
+
+  // Phase 2 members
+  const int m_log_num_segments;
+  const int m_num_bm_segments; // Number of segments in a BM - to maximize parallel operation from the serial bm sums.
+  const int m_segment_size;
+
+  /**
+   * @struct BmSumSegment
+   * @brief Struct bundling the required data members for each BM segment in phase 2.
+   */
+  struct BmSumSegment {
+    // Imagining the required BM sum as a right angle triangle - Nth element is a N high column of the element -
+    // A method to summing the BM serially is starting from the top with a triangle sum up to the current line and a
+    // line sum which is calculated for each line then added to the triangle.
+    // Therefore both sums are need as a part of this struct.
+    P triangle_sum;
+    P line_sum;
+
+    int m_nof_received_sums = 1; // Counter for numbers of sum received in this row idx - used to determine when a new
+                                 // row idx can be dispatched (Due to it depending on the two sums from the prev idx).
+    int m_idx_in_segment;        // Counter counting down how far the current sum is through the segment.
+    int m_segment_mem_start;     // Offset of segment start in memory / vector.
+  };
+
+  // Phase 3 members
+  std::vector<std::thread> m_p3_threads;
+
+  /**
+   * @brief Phase 1 (accumulation) of MSM
+   * @param scalars - scalar input.
+   * @param bases - EC P input, affine representation (Regardless of the defined P type of the class).
+   * @param msm_size - Size of the above arrays, as they are given as Pers.
+   */
+  void bucket_accumulator(const scalar_t* scalars, const A* bases, const unsigned int msm_size);
+
+  /**
+   * @brief Push addition task during phase 1.
+   * Th function also handles completed addition results while attempting to insert the new addition task (including
+   * potential collision that requires more urgent ec-addition of the result and the current stored value).
+   * @param task_bkt_idx - address in m_buckets in which to store the result in the future.
+   * @param bkt - the P from m_buckets.
+   * @param base - the P from the input bases
+   * @param negate_base - flag to signal the task to subtract base instead of adding it.
+   */
+  void phase1_push_addition(const unsigned int task_bkt_idx, const P bkt, const A base);
+  /**
+   * @brief Handles the final results of phase 1 (after no other planned additions are required).
+   * The function also handles the potential collision similarly to push_addition above.
+   */
+  void phase1_wait_for_completion();
+
+  /**
+   * @brief Phase 2 of MSM. Function handling the initial parallel part of summing the bucket modules. It splits the
+   * BMs into segments, summing each separately and passing the segments sum to be handled by the final accumulator.
+   * @return vector containing segments line and triangle sums for the final accumulator.
+   */
+  void bm_sum(std::shared_ptr<std::vector<BmSumSegment>>& segments_ptr);
```

Comment:
(1) why pass shared_ptr by reference?
(2) why do you even need a shared_ptr here? you could just have the vector and pass it by reference instead of passing the shared_ptr by reference.

Reply:
Using segments_ptr.get()?